File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null) {
            return false;
        }
        
        // Find mid point
        ListNode slow = head;
        ListNode fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode midPoint = slow;
        
        // Reverse the nodes starting from mid point
        ListNode prev = null;
        ListNode next = null;
        while (midPoint != null) {
            next = midPoint.next;
            midPoint.next = prev;
            prev = midPoint;
            midPoint = next;
        }
        
        // Check for palindrome
        ListNode left = head;
        ListNode right = prev;
        while (left != null && right != null) {
            if (left.val != right.val) {
                return false;
            }
            
            left = left.next;
            right = right.next;
        }
        
        return true;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # Find the middle node
        slow = head
        fast = head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        mid = slow
        
        # Reverse the nodes starting from the middle node
        prev = None
        while mid:
            next = mid.next
            mid.next = prev
            prev = mid
            mid = next
        
        head2 = prev
        while head2:
            if head.val != head2.val:
                return False
            head2 = head2.next
            head = head.next
        
        return True